HDDS-3947: Sort DNs for client when the key is a file for #getFileStatus #listStatus APIs

rakeshadr opened the pull request:
## What changes were proposed in this pull request?

Similar to OzoneManagerFS#lookupFile(OmKeyArgs args, String clientAddress) interface, "clientAddress" has been passed as an argument to OzoneManagerFS#listStatus(), OzoneManagerFS#getFileStatus() APIs.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-3947

Please replace this section with the link to the Apache JIRA)

## How was this patch tested?

Used existing TestKeyManagerImpl unit test cases.

adoroszlai commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/fs/OzoneManagerFS.java:
```diff
@@ -49,6 +69,21 @@ OpenKeySession createFile(OmKeyArgs args, boolean isOverWrite,
+   */
   List<OzoneFileStatus> listStatus(OmKeyArgs keyArgs, boolean recursive,
-      String startKey, long numEntries) throws IOException;
+      String startKey, long numEntries, String clientAddress)
```
In my [comment](https://github.com/apache/hadoop-ozone/pull/1192#discussion_r481025587) on the previous PR:

> Should we also keep the two original methods?

I meant both `getFileStatus` and `listStatus`.  Sorry if it was unclear.

rakeshadr replied:
Sure, I will make the changes in my next commit.

rakeshadr replied:
Sorry for re-opening the discussion on this comment. On a second thought `listStatus` is only used in [one place in the source code](https://github.com/apache/hadoop-ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java#L2910) and there won't be any other caller for `listStatus() without ClientAddress`. So, how about deleting the original method and can add it later if there is any need ?

adoroszlai replied:
> listStatus is only used in one place in the source code

It is also used by several tests, which all pass `null` as address.

rakeshadr replied:
OK, addressed in my new commit

adoroszlai reviewed (commented):
Thanks @rakeshadr for updating the patch.

rakeshadr commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java:
```diff
@@ -1783,7 +1802,9 @@ private OzoneFileStatus getOzoneFileStatus(String volumeName,
       // if the key is a file then do refresh pipeline info in OM by asking SCM
       if (fileKeyInfo != null) {
-        refreshPipeline(fileKeyInfo);
+        if (refreshPipeline) {
```
Thanks @adoroszlai again for the review help.

Yes, you are correct. Test passed locally without the `if (refreshPipeline)` condition.

Point-1) That means, the caller should call`refreshPipeline(OmKeyInfo value)` method without the `refreshPipeline` condition.

Point-2) Also, I'm planning to move the [refreshPipeline](https://github.com/apache/hadoop-ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java#L2131) call outside BUCKET_LOCK and will be moving to [here](https://github.com/apache/hadoop-ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java#L2147). Now, its a duplicate call added as part of HDDS-3658.

Does these two changes make sense to you?

adoroszlai replied:
Yes, both points make sense.  Thanks for noticing the duplicate call.

rakeshadr replied:
OK, Done!

rakeshadr commented:
Thanks @adoroszlai for the reviews. Uploaded new commit addressing your comments.

xiaoyuyao reviewed (approved):
Thanks @rakeshadr  for working on this. The patch LGTM, +1.

adoroszlai reviewed (commented):
Thanks @rakeshadr for updating the patch.

I have one question left: shouldn't `getFileStatus` call from `listStatus` pass `clientAddress`?

https://github.com/apache/hadoop-ozone/blob/088e40d2f79ab0e8c6021523fd4e2b8db23127af/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java#L2076

rakeshadr commented:
> Thanks @rakeshadr for updating the patch.
> 
> I have one question left: shouldn't `getFileStatus` call from `listStatus` pass `clientAddress`?
> 
> https://github.com/apache/hadoop-ozone/blob/088e40d2f79ab0e8c6021523fd4e2b8db23127af/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java#L2076

Good catch. Thanks @adoroszlai. 

Updated new commit. I've added clientAddress to the #getFileStatus call. Also, I've moved #getFileStatus call outside BUCKET_LOCK so that it will get the benefit of doing 'sort' and 'refreshPipeline' outside BUCKET_LOCK.

rakeshadr commented:
> Thanks @rakeshadr for working on this. The patch LGTM, +1.

Thanks a lot @xiaoyuyao for the reviews!

adoroszlai reviewed (approved):
Thanks @rakeshadr for improving this patch several times.

adoroszlai commented:
Thanks @rakeshadr for the contribution, and @xiaoyuyao for the review.

rakeshadr commented:
Thanks a lot @adoroszlai for review help and merging it.
